interfaces/builtin/broadcom-asic-control.go

```diff
@@ -0,0 +1,150 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/interfaces/apparmor"
+	"github.com/snapcore/snapd/interfaces/kmod"
+	"github.com/snapcore/snapd/interfaces/udev"
+)
+
+const broadcomAsicControlSummary = `allows using the broadcom-asic kernel module`
+
+const broadcomAsicControlBaseDeclarationSlots = `
+  broadcom-asic-control:
+    allow-installation:
+      slot-snap-type:
+        - core
+    deny-auto-connection: true
+`
+const broadcomAsicControlConnectedPlugAppArmor = `
+# Description: Allow access to broadcom asic kernel module.
+
+/sys/module/linux_kernel_bde/initstate r,
+/sys/module/linux_user_bde/initstate r,
+/sys/module/linux_kernel_bde/holders/ r,
+/sys/module/linux_user_bde/holders/ r,
+/sys/module/linux_user_bde/holders/** r,
+/sys/module/linux_user_bde/refcnt r,
+/sys/module/linux_bcm_knet/initstate r,
+/sys/module/linux_bcm_knet/holders/ r,
+/sys/module/linux_bcm_knet/refcnt r,
+/dev/linux-user-bde rw,
+/dev/linux-kernel-bde rw,
+/dev/linux-bcm-knet wr,
+`
+
+const broadcomAsicControlConnectedPlugUDev = `
+KERNEL=="linux-user-bde", TAG+="###SLOT_SECURITY_TAGS###"
+KERNEL=="linux-kernel-bde", TAG+="###SLOT_SECURITY_TAGS###"
+KERNEL=="linux-bcm-knet", TAG+="###SLOT_SECURITY_TAGS###"
+`
+
+// The upstream linux kernel doesn't come with support for the
+// necessary kernel modules we need to drive a Broadcom ASIC.
+// All necessary modules need to be loaded on demand if the
+// kernel the device runs with provides them.
+var broadcomAsicControlConnectedPlugKMod = []string{
+	"linux-user-bde",
+	"linux-kernel-bde",
+	"linux-bcm-knet",
+}
+
+// The type for broadcom-asic-control interface
+type broadcomAsicControlInterface struct{}
+
+// Getter for the name of the broadcom-asic-control interface
+func (iface *broadcomAsicControlInterface) Name() string {
+	return "broadcom-asic-control"
+}
+
+func (iface *broadcomAsicControlInterface) MetaData() interfaces.MetaData {
+	return interfaces.MetaData{
+		Summary:              broadcomAsicControlSummary,
+		ImplicitOnCore:       true,
+		ImplicitOnClassic:    true,
+		BaseDeclarationSlots: broadcomAsicControlBaseDeclarationSlots,
+	}
+}
+
+func (iface *broadcomAsicControlInterface) String() string {
+	return iface.Name()
+}
+
+// Check validity of the defined slot
+func (iface *broadcomAsicControlInterface) SanitizeSlot(slot *interfaces.Slot) error {
+	// Does it have right type?
+	if iface.Name() != slot.Interface {
+		panic(fmt.Sprintf("slot is not of interface %q", iface))
+	}
+
+	// Creation of the slot of this type is allowed only by a gadget or os snap
+	if !(slot.Snap.Type == "os") {
```

Comment:
The type of a core snap is still "os". See https://github.com/snapcore/snapd/blob/master/snap/types.go#L31 for details